nuto/mechanics/cell/SimpleAssembler.cpp

```diff
@@ -36,85 +36,115 @@ std::vector<DofType> DofIntersection(std::vector<DofType> one, std::vector<DofTy
+                                                     CellInterface::MatrixFunction f) const
 {
     ThrowOnZeroDofNumbering(dofTypes);
 
-    using TripletList = std::vector<Eigen::Triplet<double>>;
+    using TripletList = std::list<Eigen::Triplet<double>>;
     DofMatrixContainer<TripletList> triplets;
 
-    for (NuTo::CellInterface& cell : cells)
+#pragma omp parallel
     {
-        const DofMatrix<double> cellHessian = cell.Integrate(f);
-        auto dofTypesToAssemble = DofIntersection(cellHessian.DofTypes(), dofTypes);
+        DofMatrixContainer<TripletList> localtriplets;
 
-        for (DofType dofI : dofTypesToAssemble)
+#pragma omp for nowait
+        for (auto cellit = cells.begin(); cellit < cells.end(); cellit++)
         {
-            Eigen::VectorXi numberingDofI = cell.DofNumbering(dofI);
-            for (DofType dofJ : dofTypesToAssemble)
-            {
-                Eigen::VectorXi numberingDofJ = cell.DofNumbering(dofJ);
-                const Eigen::MatrixXd& cellHessianDof = cellHessian(dofI, dofJ);
+            const DofMatrix<double> cellHessian = cellit->Integrate(f);
+            auto dofTypesToAssemble = DofIntersection(cellHessian.DofTypes(), dofTypes);
 
-                for (int i = 0; i < numberingDofI.rows(); ++i)
+            for (DofType dofI : dofTypesToAssemble)
+            {
+                Eigen::VectorXi numberingDofI = cellit->DofNumbering(dofI);
+                for (DofType dofJ : dofTypesToAssemble)
                 {
-                    for (int j = 0; j < numberingDofJ.rows(); ++j)
-                    {
-                        const int globalDofNumberI = numberingDofI[i];
-                        const int globalDofNumberJ = numberingDofJ[j];
-                        const double globalDofValue = cellHessianDof(i, j);
+                    Eigen::VectorXi numberingDofJ = cellit->DofNumbering(dofJ);
+                    const Eigen::MatrixXd& cellHessianDof = cellHessian(dofI, dofJ);
 
-                        triplets(dofI, dofJ).push_back({globalDofNumberI, globalDofNumberJ, globalDofValue});
+                    for (int i = 0; i < numberingDofI.rows(); ++i)
+                    {
+                        for (int j = 0; j < numberingDofJ.rows(); ++j)
+                        {
+                            const int globalDofNumberI = numberingDofI[i];
+                            const int globalDofNumberJ = numberingDofJ[j];
+                            const double globalDofValue = cellHessianDof(i, j);
+
+                            localtriplets(dofI, dofJ).push_back({globalDofNumberI, globalDofNumberJ, globalDofValue});
+                        }
                     }
                 }
             }
         }
+#pragma omp critical
+        {
+            for (DofType dofI : dofTypes)
+                for (DofType dofJ : dofTypes)
+                {
+                    triplets(dofI, dofJ).splice(triplets(dofI, dofJ).end(), localtriplets(dofI, dofJ));
```

Comment:
I almost always prefer vector over everything. I see your point. O(1) merging of all the thread local containers. But you pay for a bad cache layout of the list. 
I would be interested in a benchmark of
a) the performance vs std::copy(vector, vector,...)
b) the actual speedup

Btw: I thought that std::deque could provide O(1) concatenation. But it seems that this is not the case.

Reply:
The bottleneck in the matrix assembly is the Eigen conversion from triplet lists to sparse matrix and since the simple assembler will vanish probably anyway I would just leave it as it is to have at least some parallel functionality (I read that also some Eigen algorithms profit from enabling omp) and do no further optimizing.

Mini benchmark of concatenation of two ranges (list/vector) with 1000 elements each using splice, insert and copy. For the vectors space is preallocated using reserve.

Benchmark          |      Time       |    CPU        |Iterations
-----------------------|-----------------|---------------|---------------
BM_ListSplice      |     58 ns       |  58 ns         | 10709471
BM_VectorInsert  |      712 ns    |    712 ns      |  968647
BM_VectorCopy  |      34952 ns |     34952 ns |     19807

Maybe not surprising.

Reply:
> Maybe not surprising.

O(1) vs O(n), yep, not surprising. On the other hand, vector allocation is O(1) or O(log n) and list allocation is O(n). Traversing a vector is faster because of cache locality. So the question is: How does the complete process of assembly react to this?

So a simulation of the conatiner based assembly could look like the last two benchmarks, for 1000 elements created twice, concatenated once and traversed twice.

Benchmark                       |            Time   |    CPU  |Iterations
-----------------------|--------------------|----------|------------------
BM_CreateVector       |                   4739 ns     |  4739 ns   |  142789
BM_CreateList             |              78747 ns    |  78746 ns    |   8126
BM_TraverseVector      |                  2323 ns   |    2323 ns  |   307932
BM_TraverseList            |              3752 ns     |  3751 ns   |  156200
BM_CreateConcatenateTraverseVector  |    23527 ns   |   23527 ns   |   30166
BM_CreateConcatenateTraverseList     |  171229 ns  |   171224 ns     |  4129

Reply:
You are right. Using std::vector will speed up the conversion from triplet lists to sparse matrix and the overall result is then faster (using insert for concatenation of vectors). I will change that.